### ISSUE TITLE
proposal: use functional options

### ISSUE THREAD

**Issue by gbbr** [opentracing/opentracing-go#174]
The current implementation has an interface `StartSpanOption` where Dave Cheney's [article](http://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) is linked specifying that it is the inspiration for this functionality.

In the article, functional options are presented which are great and are used in many libraries. In functional options we have functions that return the same type, automatically grouping them nicely in documentation, such as for example in the [grpc](https://godoc.org/google.golang.org/grpc) package.

In this library however, they are defined as an interface for no apparent reason. In this case, not only are they harder to implement and provide no real benefit, but also they do not group together in documentation, making doc pages clumsy and hard to follow.

I'd like to suggest to change this.

**Comment by yurishkuro** [comment]
NB: `StartSpanOptions` should be a private type.
